shared/src/main/kotlin/org/javacs/kt/classpath/GradleClassPathResolver.kt

```diff
@@ -19,15 +19,18 @@ internal class GradleClassPathResolver(private val path: Path, private val inclu
 
```

Comment:
Rather than building string commands, invoking the build and parsing the string output, line by line, I think that this whole file could just become a single call to the classpath, using the [Eclipse model](https://docs.gradle.org/current/javadoc/org/gradle/tooling/model/eclipse/EclipseProject.html) from the [Gradle project connection](https://docs.gradle.org/current/javadoc/org/gradle/tooling/ProjectConnection.html#notifyDaemonsAboutChangedPaths-java.util.List-). There's more information on the [third party tooling web doc](https://docs.gradle.org/current/userguide/third_party_integration.html#embedding). They say that to find examples of calling Gradle's API from Java, you need to go through the JavaDoc, starting at the [GradleConnector](https://docs.gradle.org/current/javadoc/org/gradle/tooling/GradleConnector.html).


```kotlin
package org.javacs.kt.classpath

import org.gradle.tooling.GradleConnector
import org.gradle.tooling.model.eclipse.EclipseProject
import java.io.File
import java.nio.file.Path

internal class GradleClassPathResolver(private val path: Path) : ClassPathResolver {
    override val resolverType: String = "Gradle"
    private val projectDirectory: Path get() = path.parent
    override val classpath: Set<ClassPathEntry> = GradleConnector
        .newConnector()
        .forProjectDirectory(File(projectDirectory.toString()))
        .connect()
        .use { connector ->
            val model = connector.getModel(EclipseProject::class.java)
            model.classpath.map {
                ClassPathEntry(
                    compiledJar = it.source.toPath(),
                    sourceJar = null,
                )
            }.toSet()
        }
}
```

Reply:
The main issue is that apparently the `EclipseProject` doesn't catch all the dependencies, especially for android dependencies. @Strum355 mentioned above that for SCIP they had to use a combination of `EclipseProject`, `IdeaProject`, and a gradle task that also builds some strings and passes them in to be parsed. I currently have a locally working branch that does it similarly, but the gradle task with string parsing definitely makes it slower.

When I have some time I'll try to make some comparisons of the outputs of each.

Reply:
Sounds great, thanks @aiguofer! Strum's comment now makes much more sense!